revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,232 @@
+// errors. The size of these dimensions is as follows:
+//   - 1st dimension will always be equal in length to the certificate chain.
+//   - 2nd will have a length between 1 and the number of OCSPServers for the cert
+//
+// At each index, nil will be used to indicate no error. If there is an error,
+// it will be one of the errors defined in errors.go.
+//
+// (e.g. for a chain with 3 certs (leaf, intermediate, root) with both the leaf
+// and intermediate specifying 2 OCSP server, this could be a possible result):
+//
+//	[[PKIXNoCheckError, TimeoutError], [RevokedError], [nil]]
+func OCSPStatus(opts Options) [][]error {
+	certResults := make([][]error, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return certResults
+	}
+
+	// Validate cert chain structure
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")})
+				return certResults
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")})
+				return certResults
+			}
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			certResults[i] = certOCSPStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return certResults
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert, issuer *x509.Certificate, opts Options) []error {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return []error{NoOCSPServerError{}}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		status, err := checkStatusFromServer(cert, issuer, server, opts)
+		if err != nil {
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		// A valid response has been received from an OCSP server
+		// Result should be based on only this response, not any errors from
+		// other servers
```

Comment:
Why? Can we keep the status of previous servers instead?

Reply:
The reason is that if we return a list of errors that ends in nil (meaning the certificate isn't revoked), the other errors from previous server attempts would overshadow the actual result. Essentially, we are only looking for one valid OCSP response. The same way that once we get the valid response we don't check other servers. The core of what we are looking for is a valid response from one of the servers, so we drop the other errors.

Reply:
Again, this can be misleading. This issue can be addressed if there is a better struct to return OCSP result.